src/http_server.c

```diff
@@ -0,0 +1,155 @@
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <string.h>
+#include <unistd.h>
+#include "config.h"
+#include "log.h"
+#include <pthread.h>
+#include "tunnel.h"
+#include <ctype.h>
+#include <signal.h>
+#include <netinet/tcp.h>
+
+
+
+int process_request(int new_socket, char *id) {
+    log_info("Processing request\n");
+    
+    int flag = 1;
+    setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+    const int buffer_size = 2048;
+    char *response[buffer_size];
+    memset(response,' ',buffer_size);
+    response[buffer_size - 1] = '\0';
+    char *reply = "HTTP/1.1 200 OK\r\n"
+            "Content-Type: text/plain\r\n"
+            "Connection: close\r\n\r\n"
+            "SAML Login...\r\n\0";
+    memcpy(response,reply,strlen(reply));
+
+    // Read the request
+    ssize_t write_result = write(new_socket, response, buffer_size);
+    (void)write_result;
+
+    // dup2(new_socket, STDOUT_FILENO);
+    // dup2(new_socket, STDERR_FILENO);
+
+
+
+    char request[1024];
+    ssize_t read_result = read(new_socket, request, sizeof(request));
+
+    // Check for '=id' in the response
+    if (read_result < 0 || strlen(request) < 10 || strncmp(request, "GET /?id=", 9) != 0) {
+        log_error("Bad request\n");
+        return -1;
+    }
+
+    // Extract the id
+    const int id_start = 9;
+    char *id_end = memchr(&request[id_start], ' ', 1000);
+
+    if (id_end == NULL) {
+        log_error("Bad request format\n");
+        return -1;
+    }
+
+    int id_length = id_end - &request[id_start];
+
+    if(id_length == 0 || id_length > 1000) {
+        log_error("Bad request id\n");
+        return -1;
+    }
+
+    strncpy(id, &request[id_start], id_length);
+
+    for (int i = 0; i < id_length; i++) {
+        if (isalnum(id[i]) || id[i] == '-') continue;
+        log_error("Invalid id format\n");
+        return -1;
+    }
+    // close(new_socket);
```

Comment:
Why not closing the socket? I can't find any other code location where this is done.